### DISCUSSION TITLE
Using membrane in place of Medialive for a live-transcoder

### DISCUSSION THREAD

**Discussion by grant-zukowski-xumo** [membraneframework/membrane-core#878] [General]
Hi, my business use case has us digesting usually an HLS stream as a source, then transcoding (usually the highest variant) into another lower quality hls variant and or an RTP connection to an outside entity.  So we may have one HLS input for example and want 2 outputs, HLS and/or RTP.  Do you think I could run membrane on an EC2 instance and achieve this goal?

**Comment by mat-hek** [comment]
Hi there, that should be possible ;) See
- https://github.com/kim-company/membrane_hls_plugin for HLS ingest
- https://github.com/membraneframework/membrane_h264_ffmpeg_plugin for transcoding
- https://github.com/membraneframework/membrane_tee_plugin for connecting multiple outputs
- https://github.com/membraneframework/membrane_http_adaptive_stream_plugin for HLS egress
- https://github.com/membraneframework/membrane_rtp_plugin for RTP egress

**Reply by grant-zukowski-xumo** [reply]
To clarify, the example for https://github.com/kim-company/membrane_hls_plugin for HLS ingest we are using the HLS as our source.  The example in this repo has the source as RTMP.  I just wanted to make sure you are saying this library can let us use HLS as source.

**Reply by grant-zukowski-xumo** [reply]
I think I answered my own question.  They are listing the rtmp out as #source, but I can see the HLS is used to ingest.

**Reply by grant-zukowski-xumo** [reply]
sorry I'm still not sure.  I'm not sure the example in the https://github.com/kim-company/membrane_hls_plugin for HLS ingest is ingesting HLS.

**Reply by mat-hek** [reply]
The HLS plugin is for HLS ingest, while HTTP Adaptive Stream is for egress :P We're going to unify that one day. For the details of the HLS plugin it's best to ask its authors ;)

**Comment by grant-zukowski-xumo** [comment]
Thank you, another question, we are concerned about CUE markers and SCTE-35 messages.  Are there libraries that let us manipulate these?

**Comment by mat-hek** [comment]
Nope, they would have to be supported by the HTTP adaptive stream plugin

**Comment by grant-zukowski-xumo** [comment]
I've been trying to work with the https://github.com/kim-company/membrane_hls_plugin for HLS ingest team on their repo, but I can't get the most basic hls input working yet.  Super frustrating experience so far.  The only example they have is for a different use of hls, mainly as an output.

https://github.com/kim-company/membrane_hls_plugin/issues/12

here is my thread over there.

**Reply by grant-zukowski-xumo** [reply]
OK, I got the source to work I think, now I'm trying to figure out how to get an output.

**Comment by grant-zukowski-xumo** [comment]
Still feeling frustrated.  I've read the introductory docs, and almost all the examples.  I still can't get even the most basic pipeline working.  

I have a source, I just want to dump the files to see if this is actually working.  Ideally I would take the highest variant, then pass that on as an hls out.  I feel like I'm just copy and pasting things and seeing what works, which is not how I like to work.  

```
 def handle_init(_ctx, _opts) do
    structure =
      child(:source, %Membrane.HLS.Source{
        reader: %My.Reader{},
        master_playlist_uri:
          "<my hls source>"
      })
      |> child(:sink, %Membrane.File.Sink{location: "thing.raw"})

    {[spec: structure], %{}}
  end
  ```
